core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala

```diff
@@ -463,39 +463,47 @@ trait EntityComponent {
       }
       else {
         DBIO.successful(Seq.empty[EntityRecord])
       }
     }
 
-    def selectEntityIds(workspaceContext: SlickWorkspaceContext, entities: Seq[EntityRecord]): ReadAction[Seq[EntityRecord]] = {
+    def selectEntityIds(workspaceContext: SlickWorkspaceContext, entities: Seq[(String, String)]): ReadAction[Seq[EntityRecord]] = {
       val entitiesGrouped = entities.grouped(batchSize).toSeq
 
       DBIO.sequence(entitiesGrouped map { batch =>
-        EntityRecordRawSqlQuery.action(workspaceContext.workspaceId, batch.map(_.toReference))
+        EntityRecordRawSqlQuery.action(workspaceContext.workspaceId, batch.map{ case (entityType, name) => AttributeEntityReference(entityType, name) })
       }).map(_.flatten)
     }
 
-    def cloneEntities(destWorkspaceContext: SlickWorkspaceContext, entities: TraversableOnce[Entity]): ReadWriteAction[Int] = {
-      batchInsertEntities(destWorkspaceContext, entities.toSeq.map(marshalNewEntity(_, destWorkspaceContext.workspaceId))) flatMap { ids =>
-        val entityIdByEntity = ids.map(record => record.id -> entities.filter(p => p.entityType == record.entityType && p.name == record.name).toSeq.head)
+    def cloneEntities(destWorkspaceContext: SlickWorkspaceContext, entitiesToClone: TraversableOnce[Entity], entitiesToReference: Seq[Entity]): ReadWriteAction[Int] = {
+      batchInsertEntities(destWorkspaceContext, entitiesToClone.toSeq.map(marshalNewEntity(_, destWorkspaceContext.workspaceId))) flatMap { ids =>
+        val entityIdByEntity = ids.map(record => record.id -> entitiesToClone.filter(p => p.entityType == record.entityType && p.name == record.name).toSeq.head)
         val entityIdsByRef = entityIdByEntity.map { case (entityId, entity) => entity.toReference -> entityId }.toMap
 
-        val attributeRecords = entityIdByEntity flatMap { case (entityId, entity) =>
-          entity.attributes.flatMap { case (attributeName, attr) =>
-            entityAttributeQuery.marshalAttribute(entityId, attributeName, attr, entityIdsByRef)
+        val entitiesToReferenceById = selectEntityIds(destWorkspaceContext, entitiesToReference.map(e => (e.entityType, e.name)))
+
+        val attributeRecords = entitiesToReferenceById map { entityRecsToReference =>
+          val entityIdsToReferenceById = entityRecsToReference.map(record => record.id -> entitiesToReference.filter(p => p.entityType == record.entityType && p.name == record.name).head)
+          val entityIdsToReferenceByRef = entityIdsToReferenceById.map { case (entityId, entity) => entity.toReference -> entityId }.toMap
+          entityIdByEntity flatMap { case (entityId, entity) =>
+            entity.attributes.flatMap { case (attributeName, attr) =>
+              entityAttributeQuery.marshalAttribute(entityId, attributeName, attr, (entityIdsByRef ++ entityIdsToReferenceByRef))
+            }
```

Comment:
I am finding this function very confusing. First I think changing entitiesToReference to be Seq[AttributeEntityReference] will be helpful. Then there are a lot of lines of code to get a map of entity ref to database id and it is mixed with marshaling attributes. I would have 1 function call to take entitiesToReference and return Map[AttributeEntityReference, Long]. Last, the nested flatMaps are a pain, try using a for comprehension instead.